src/python/espressomd/particle_data.pyx

```diff
@@ -1556,14 +1556,43 @@ cdef class _ParticleSliceImpl:
 
     def __cinit__(self, slice_, prefetch_chunk_size=10000):
         self._chunk_size = prefetch_chunk_size
-
-        id_list = np.arange(max_seen_particle + 1)
-        self.id_selection = id_list[slice_]
-        mask = np.empty(len(self.id_selection), dtype=np.bool)
-        cdef int i
-        for i in range(len(self.id_selection) - 1, -1, -1):
-            mask[i] = particle_exists(i)
-        self.id_selection = self.id_selection[mask]
+        # We distinguish two cases:
+        # * ranges and slices only specify lower and upper bounds for particle
+        #   ids and optionally a step. Gaps in the id range are tolerated.
+        # * Explicit list/tuple/ndarrya containing particle ids. Here
+        #   all particles have to exist and the result maintains the order
+        #   specified in the list.
+        if isinstance(slice_, (slice, range)):
+            # Prevent negavite bounds
+            if (not slice_.start is None and slice_.start < 0) or\
+               (not slice_.stop is None and slice_.stop < 0):
+                raise IndexError(
+                    "Negative start and end ids are not supported on ParticleSlice")
+            id_list = np.arange(max_seen_particle + 1, dtype=int)
+            self.id_selection = id_list[slice_]
+
+            # Remove non-existing particles form the id selection
+            mask = np.empty(len(self.id_selection), dtype=np.bool)
+            mask[:] = True
+            for i, id in enumerate(self.id_selection):
+                if not particle_exists(id):
+                    mask[i] = False
+            self.id_selection = self.id_selection[mask]
+        elif isinstance(slice_, (list, tuple, np.ndarray)):
+            # Check that all entries are some flavor of integer
+            for id in slice_:
+                if not is_valid_type(id, int):
+                    raise TypeError(
+                        "Particle id must be an integer but got " + str(id))
+                if not particle_exists(id):
+                    raise IndexError("Particle does not exist " + str(id))
+            self.id_selection = np.array(slice_, dtype=int)
```

Comment:
I think all this code does not belong here. We should put the slice validation into some dedicated place like ```utils```. It should also be tested.

Reply:
Moved to individaul member functoins in ParticleSlice